File: app/src/main/java/ui/fragment/AddEditTodoFragment.java
package ui.fragment;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Typeface;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.transition.Slide;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.IdRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.core.content.ContextCompat;
import androidx.core.widget.NestedScrollView;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;
import androidx.lifecycle.ViewModelProvider;

import com.google.android.material.button.MaterialButton;
import com.google.android.material.card.MaterialCardView;
import com.google.android.material.chip.ChipGroup;
import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textfield.TextInputLayout;

import org.jetbrains.annotations.NotNull;

import java.text.MessageFormat;
import java.util.List;
import java.util.Objects;

import hlv.cute.todo.R;
import hlv.cute.todo.databinding.FragmentAddEditTodoBinding;
import ir.hamsaa.persiandatepicker.PersianDatePickerDialog;
import ir.hamsaa.persiandatepicker.api.PersianPickerDate;
import ir.hamsaa.persiandatepicker.api.PersianPickerListener;
import model.Category;
import model.Todo;
import scheduler.alarm.AlarmUtil;
import ui.dialog.DropDownCategoriesDialog;
import ui.dialog.TimePickerSheetDialog;
import utils.Constants;
import utils.ResourceUtils;
import utils.ToastHelper;
import viewmodel.AddEditTodoViewModel;

public class AddEditTodoFragment extends BaseFragment {

    private static final String TODO_ARGS = "todo-args";
    private static final String SHARE_MODE_ARGS = "share-mode-args";

    private FragmentAddEditTodoBinding binding;

    private AddEditTodoViewModel viewModel;

    private MaterialButton btnAdd;
    private TextInputLayout inpLytTitle;
    private TextInputEditText edtTitle;
    private ChipGroup chipGP;
    private ConstraintLayout toolbar;
    private NestedScrollView nested;
    private TextView txtTitle;
    private MaterialCardView cardCategory;
    private MaterialCardView cardReminder;
    private AppCompatImageView imgClear;
    private TextView txtCategory;
    private TextView txtDate;

    public AddEditTodoFragment() {
    }

    public static AddEditTodoFragment newInstance(Todo todo) {
        AddEditTodoFragment fragment = new AddEditTodoFragment();
        Bundle args = new Bundle();
        args.putSerializable(TODO_ARGS, todo);
        fragment.setArguments(args);
        return fragment;
    }

    public static AddEditTodoFragment newInstanceShare(String title) {
        AddEditTodoFragment fragment = new AddEditTodoFragment();
        Bundle args = new Bundle();
        args.putString(SHARE_MODE_ARGS, title);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        viewModel = new ViewModelProvider(this).get(AddEditTodoViewModel.class);

        if (getArguments() != null && !getArguments().isEmpty()) {
            Todo todo = (Todo) getArguments().getSerializable(TODO_ARGS);

            if (todo != null) {
                //edit mode
                viewModel.setEditMode(true);
                viewModel.setShareMode(false);
                viewModel.setTodo(todo);
            } else {
                //add mode (share or normal)
                viewModel.setEditMode(false);

                String shareTitle = getArguments().getString(SHARE_MODE_ARGS);
                if (shareTitle != null && !shareTitle.isEmpty()) {
                    viewModel.setShareMode(true);
                    viewModel.setShareTitle(shareTitle);
                } else {
                    viewModel.setShareMode(false);
                }
            }
        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentAddEditTodoBinding.inflate(inflater);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initViews();
        initLogic();
        handleAction();
        handleObserver();
    }

    private void initViews() {
        binding.aImgBack.setOnClickListener(view -> {
            if (viewModel.isShareMode()) {
                back();
                Fragment homeFragment = HomeFragment.newInstance();
                FragmentTransaction transaction = getParentFragmentManager().beginTransaction();
                transaction.add(R.id.mainContainer, homeFragment, Constants.FragmentTag.HOME);
                transaction.commit();
                return;
            }

            back();
        });

        inpLytTitle = binding.inpTitle;
        edtTitle = binding.inpEdtTitle;
        btnAdd = binding.mBtnAdd;
        chipGP = binding.chipGP;
        toolbar = binding.toolbar;
        nested = binding.nested;
        txtTitle = binding.txtTitle;
        cardCategory = binding.mCardCategory;
        cardReminder = binding.mCardReminder;
        imgClear = binding.aImgClear;
        txtCategory = binding.txtCategory;
        txtDate = binding.txtDate;

        handleScroll();
    }

    @SuppressLint("ClickableViewAccessibility")
    private void handleScroll() {
        edtTitle.setOnTouchListener((v, event) -> {
            if (edtTitle.hasFocus()) {
                v.getParent().requestDisallowInterceptTouchEvent(true);
                if ((event.getAction() & MotionEvent.ACTION_MASK) == MotionEvent.ACTION_SCROLL) {
                    v.getParent().requestDisallowInterceptTouchEvent(false);
                    return true;
                }
            }
            return false;
        });


        nested.setOnScrollChangeListener(new NestedScrollView.OnScrollChangeListener() {
            final float dpShadow = getResources().getDimension(R.dimen.toolbar_shadow);

            @Override
            public void onScrollChange(NestedScrollView v, int scrollX, int scrollY, int oldScrollX, int oldScrollY) {
                if (scrollY == 0) {
                    toolbar.animate().translationZ(0).setStartDelay(0).setDuration(200).start();
                    //toolbar.setTranslationZ(0);
                } else if (scrollY > 0) {
                    toolbar.setTranslationZ(dpShadow);
                    toolbar.animate().translationZ(dpShadow).setStartDelay(0).setDuration(90).start();
                }
            }
        });
    }

    @SuppressLint("NonConstantResourceId")
    private void initLogic() {
        txtTitle.setText(viewModel.getTitleFragment());
        edtTitle.setText(viewModel.getEditTextTitle());

        btnAdd.setText(viewModel.getButtonPrimaryText());

        viewModel.firstInitDateTime();

        if (viewModel.isEditMode()) {
            @IdRes int chipID;
            switch (viewModel.getTodo().getPriority()) {
                case LOW:
                default:
                    chipID = R.id.chipLow;
                    viewModel.setPriority(Todo.Priority.LOW);
                    break;

                case NORMAL:
                    chipID = R.id.chipNormal;
                    viewModel.setPriority(Todo.Priority.NORMAL);
                    break;

                case HIGH:
                    chipID = R.id.chipHigh;
                    viewModel.setPriority(Todo.Priority.HIGH);
                    break;
            }

            chipGP.check(chipID);
            return;
        }

        //set default priority
        chipGP.check(R.id.chipLow);
        viewModel.setPriority(Todo.Priority.LOW);
    }

    @SuppressLint("NonConstantResourceId")
    private void handleAction() {
        cardCategory.setOnClickListener(view -> {
            List<Category> categories = getCategoryViewModel().getAllCategories();

            DropDownCategoriesDialog dropDown = new DropDownCategoriesDialog(getActivity(), categories);
            dropDown.show();

            dropDown.setOnClickCategory(category -> {
                dropDown.dismiss();
                viewModel.commitCategory(category);
            });

            dropDown.setOnclickManage(() -> {
                viewModel.commitCategory(null);

                Fragment fragment = CategoriesFragment.newInstance();
                fragment.setEnterTransition(new Slide(Gravity.BOTTOM));
                FragmentTransaction transaction = getParentFragmentManager().beginTransaction();
                transaction.add(R.id.mainContainer, fragment, Constants.FragmentTag.CATEGORY);
                transaction.addToBackStack(Constants.BackStack.CATEGORY);
                transaction.commit();

                dropDown.dismiss();
            });

        });

        chipGP.setOnCheckedChangeListener((group, checkedId) -> {
            switch (checkedId) {
                case R.id.chipLow:
                default:
                    viewModel.setPriority(Todo.Priority.LOW);
                    break;

                case R.id.chipNormal:
                    viewModel.setPriority(Todo.Priority.NORMAL);
                    break;

                case R.id.chipHigh:
                    viewModel.setPriority(Todo.Priority.HIGH);
                    break;
            }
        });

        edtTitle.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void afterTextChanged(Editable editable) {
                if (editable.length() > 0)
                    inpLytTitle.setError(null);
            }
        });

        cardReminder.setOnClickListener(view -> handlePickers(getActivity(), viewModel.getOldDateTime().getDate()));

        imgClear.setOnClickListener(view -> viewModel.releaseAll());

        btnAdd.setOnClickListener(view -> {
            inpLytTitle.setError(null);

            if (viewModel.isEditMode()) {
                Todo editedTodo = viewModel.editTodo(Objects.requireNonNull(edtTitle.getText()).toString().trim());

                String res = getTodoViewModel().validateTodo(editedTodo);
                if (res == null) {
                    if (editedTodo.getArriveDate() != 0) {
                        AlarmUtil.get().cancelAlarm(editedTodo.getId());
                        AlarmUtil.get().setAlarm(editedTodo.getId(), editedTodo.getTitle(), editedTodo.getArriveDate());
                    }

                    getTodoViewModel().editTodo(editedTodo);
                    getSearchViewModel().fetch();
                    updateDetail(editedTodo);
                    back();
                    return;
                }

                inpLytTitle.setError(res);
            } else {
                Todo newTodo = viewModel.addTodo(Objects.requireNonNull(edtTitle.getText()).toString().trim());

                String res = getTodoViewModel().validateTodo(newTodo);
                if (res == null) {
                    if (newTodo.getArriveDate() != 0)
                        AlarmUtil.get().setAlarm(newTodo.getId(), newTodo.getTitle(), newTodo.getArriveDate());

                    getTodoViewModel().goToTop();
                    getTodoViewModel().addTodo(newTodo);
                    ToastHelper.get().toast(getString(R.string.todo_added_successfully));
                    back();

                    if (viewModel.isShareMode()) {
                        btnAdd.setEnabled(false);

                        Fragment homeFragment = HomeFragment.newInstance();
                        FragmentTransaction transaction = getParentFragmentManager().beginTransaction();
                        transaction.add(R.id.mainContainer, homeFragment, Constants.FragmentTag.HOME);
                        transaction.commit();
                    }

                    return;
                }

                inpLytTitle.setError(res);
            }
        });
    }

    private void handlePickers(Context context, PersianPickerDate persianDate) {
        PersianDatePickerDialog picker = new PersianDatePickerDialog(context)
                .setPositiveButtonString("مرحله بعد")
                .setNegativeButton("انصراف")
                .setTodayButton("تاریخ امروز")
                .setTodayButtonVisible(true)
                .setMinYear(1400)
                .setMaxYear(1440)
                .setTypeFace(Typeface.createFromAsset(context.getAssets(), "font/vazir_medium.ttf"))
                .setTitleColor(ContextCompat.getColor(context, R.color.blue))
                .setAllButtonsTextSize(15)
                .setPickerBackgroundColor(ContextCompat.getColor(context, R.color.white))
                .setTitleType(PersianDatePickerDialog.WEEKDAY_DAY_MONTH_YEAR)
                .setShowInBottomSheet(true)
                .setListener(new PersianPickerListener() {
                    @Override
                    public void onDateSelected(@NotNull PersianPickerDate persianPickerDate) {
                        viewModel.configTempDateTime();
                        viewModel.setDateTemp(persianPickerDate);

                        TimePickerSheetDialog sheetTimer = new TimePickerSheetDialog(context, viewModel.getTempDateTime());
                        sheetTimer.show();

                        sheetTimer.setOnClickApply(pickedDateTime -> {
                            viewModel.commitDateTime(pickedDateTime);
                            viewModel.commitOldDateTime(pickedDateTime);
                            sheetTimer.dismiss();
                        });

                        sheetTimer.setOnBackClick(date -> handlePickers(context, date));
                    }

                    @Override
                    public void onDismissed() {
                    }
                });

        //init default values
        int[] date = viewModel.setInitDateValue(persianDate);
        int year = date[0];
        int month = date[1];
        int day = date[2];

        picker.setInitDate(year, month, day);

        picker.show();

    }

    private void handleObserver() {
        viewModel.getDateTimeLiveData().observe(getViewLifecycleOwner(), changedDateTime -> {
            if (changedDateTime != null && changedDateTime.getDate() != null) {
                txtDate.setTextColor(ContextCompat.getColor(txtDate.getContext(), R.color.black));
                txtDate.setText(MessageFormat.format("{0}\nساعت {1}",
                        changedDateTime.getPersianDate(),
                        changedDateTime.getClock())
                );
                imgClear.setVisibility(View.VISIBLE);
            } else {
                txtDate.setTextColor(ContextCompat.getColor(txtDate.getContext(), R.color.gray));
                txtDate.setText(getString(R.string.set_date_time));
                imgClear.setVisibility(View.GONE);
            }
        });

        viewModel.getCategoryLiveData().observe(getViewLifecycleOwner(), category -> {
            txtCategory.setText(viewModel.getCategoryTitleText());

            if (viewModel.categoryIsValid())
                txtCategory.setTextColor(ResourceUtils.get().getColor(R.color.black));
            else
                txtCategory.setTextColor(ResourceUtils.get().getColor(R.color.gray));
        });
    }

    private void updateDetail(Todo todo) {
        TodoDetailFragment fragment =
                (TodoDetailFragment) getParentFragmentManager().findFragmentByTag(Constants.FragmentTag.TODO_DETAIL);

        if (fragment != null) {
            if (fragment.getViewModel() != null)
                fragment.getViewModel().setTodo(todo);
        }
    }

    public AddEditTodoViewModel getViewModel() {
        return viewModel;
    }
}

File: app/src/main/java/ui/fragment/AddEditTodoFragment.kt
package ui.fragment

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Typeface
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.transition.Slide
import android.view.*
import androidx.annotation.IdRes
import androidx.core.os.bundleOf
import androidx.core.widget.NestedScrollView
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.lifecycleScope
import dagger.hilt.android.AndroidEntryPoint
import dagger.hilt.android.qualifiers.ActivityContext
import hlv.cute.todo.R
import hlv.cute.todo.databinding.FragmentAddEditTodoBinding
import ir.hamsaa.persiandatepicker.PersianDatePickerDialog
import ir.hamsaa.persiandatepicker.api.PersianPickerDate
import ir.hamsaa.persiandatepicker.api.PersianPickerListener
import kotlinx.coroutines.launch
import model.Category
import model.DateTime
import model.Priority
import model.Todo
import ui.component.bindingComponent.BaseViewBindingFragment
import ui.dialog.DropDownCategoriesDialog
import ui.dialog.ReminderGuideDialog
import ui.dialog.TimePickerSheetDialog
import ui.dialog.WarningDateDialog
import utils.Constants
import utils.ToastUtil
import utils.collectLatestLifecycleFlow
import utils.collectLifecycleFlow
import viewmodel.AddEditTodoViewModel
import viewmodel.NotificationViewModel
import java.text.MessageFormat
import javax.inject.Inject

@AndroidEntryPoint
class AddEditTodoFragment : BaseViewBindingFragment<FragmentAddEditTodoBinding>() {

    override val bindingInflater: (LayoutInflater, ViewGroup?, Boolean) -> FragmentAddEditTodoBinding
        get() = FragmentAddEditTodoBinding::inflate

    val viewModel by viewModels<AddEditTodoViewModel>()
    private val notificationViewModel by viewModels<NotificationViewModel>()

    @Inject
    lateinit var toastUtil: ToastUtil

    @Inject
    @ActivityContext
    lateinit var iContext: Context

    private var allCategories: ArrayList<Category>? = null

    companion object {
        private const val TODO_ARGS = "todo-args"
        private const val SHARE_MODE_ARGS = "share-mode-args"

        @JvmStatic
        fun newInstance(todo: Todo?): AddEditTodoFragment {
            val fragment = AddEditTodoFragment()

            val args = bundleOf(
                TODO_ARGS to todo
            )

            fragment.arguments = args

            return fragment
        }

        @JvmStatic
        fun newInstanceShare(title: String?): AddEditTodoFragment {
            val fragment = AddEditTodoFragment()

            val args = bundleOf(
                SHARE_MODE_ARGS to title
            )

            fragment.arguments = args

            return fragment
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        if (arguments != null && arguments!!.isEmpty.not()) {

            val todo: Todo? = arguments!!.getParcelable(TODO_ARGS)

            if (todo != null) {
                //edit mode
                viewModel.isEditMode = true
                viewModel.isShareMode = false
                viewModel.todo = todo
            } else {
                //add mode (share or normal)
                viewModel.isEditMode = false
                val shareTitle = arguments!!.getString(SHARE_MODE_ARGS)

                if (shareTitle != null) {
                    viewModel.isShareMode = true
                    viewModel.shareTitle = shareTitle
                } else {
                    viewModel.isShareMode = false
                }

            }
        }
    }

    override fun initiate() {
        handleViews()
        initLogic()
        handleAction()
        handleObserver()
    }

    private fun handleViews() {
        binding.aImgBack.setOnClickListener {
            if (viewModel.isShareMode) {
                back()

                val homeFragment: Fragment = HomeFragment.newInstance()

                parentFragmentManager.beginTransaction().apply {
                    add(R.id.mainContainer, homeFragment, Constants.FragmentTag.HOME)
                }.commit()

                return@setOnClickListener
            }

            back()
        }

        binding.aImgReminderGuide.setOnClickListener {
            ReminderGuideDialog(iContext).apply {
                show()

                setTitle(provideResource.getString(R.string.set_reminder))

                setMessage(provideResource.getString(R.string.reminder_guide))
            }
        }

        handleScroll()
    }

    @SuppressLint("ClickableViewAccessibility")
    private fun handleScroll() {
        binding.inpEdtTitle.setOnTouchListener { v: View, event: MotionEvent ->
            // FIXME:
            if (binding.inpEdtTitle.hasFocus()) {
                v.parent.requestDisallowInterceptTouchEvent(true)
                if (event.action and MotionEvent.ACTION_MASK == MotionEvent.ACTION_SCROLL) {
                    v.parent.requestDisallowInterceptTouchEvent(false)
                    return@setOnTouchListener true
                }
            }

            false
        }


        binding.nested.setOnScrollChangeListener(object : NestedScrollView.OnScrollChangeListener {
            val dpShadow = provideResource.getDimen(R.dimen.toolbar_shadow)
            override fun onScrollChange(
                v: NestedScrollView,
                scrollX: Int,
                scrollY: Int,
                oldScrollX: Int,
                oldScrollY: Int
            ) {
                if (scrollY == 0) {
                    binding.toolbar.animate()
                        .translationZ(0f)
                        .setStartDelay(0)
                        .setDuration(200)
                        .start()

                    //toolbar.setTranslationZ(0);
                } else if (scrollY > 0) {
                    binding.toolbar.translationZ = dpShadow

                    binding.toolbar.animate()
                        .translationZ(dpShadow)
                        .setStartDelay(0)
                        .setDuration(90)
                        .start()
                }
            }
        })
    }

    @SuppressLint("NonConstantResourceId")
    private fun initLogic() {
        binding.txtTitle.text = viewModel.titleFragment

        binding.inpEdtTitle.setText(viewModel.editTextTitle)

        binding.mBtnAdd.text = viewModel.buttonPrimaryText

        viewModel.firstInitDateTime()

        if (viewModel.isEditMode) {

            @IdRes val chipID: Int
            when (viewModel.todo?.priority) {
                Priority.LOW -> {
                    chipID = R.id.chipLow
                    viewModel.priority = Priority.LOW
                }

                Priority.NORMAL -> {
                    chipID = R.id.chipNormal
                    viewModel.priority = Priority.NORMAL
                }

                Priority.HIGH -> {
                    chipID = R.id.chipHigh
                    viewModel.priority = Priority.HIGH
                }

                else -> {
                    chipID = R.id.chipLow
                    viewModel.priority = Priority.LOW
                }
            }

            binding.chipGP.check(chipID)
            return
        }

        //set default priority
        binding.chipGP.check(R.id.chipLow)
        viewModel.priority = Priority.LOW
    }

    @SuppressLint("NonConstantResourceId")
    private fun handleAction() {
        binding.mCardCategory.setOnClickListener {
            DropDownCategoriesDialog(activity!!, allCategories).apply {
                show()

                onClickCategory = { category: Category? ->
                    dismiss()
                    viewModel.commitCategory(category)
                }

                onclickManage = {
                    viewModel.commitCategory(null)

                    val fragment = CategoriesFragment.newInstance().apply {
                        enterTransition = Slide(Gravity.BOTTOM)
                    }

                    parentFragmentManager.beginTransaction().apply {
                        add(R.id.mainContainer, fragment, Constants.FragmentTag.CATEGORY)

                        addToBackStack(Constants.FragmentTag.CATEGORY)
                    }.commit()

                    dismiss()
                }
            }
        }

        /*chipGP!!.setOnCheckedChangeListener { group: ChipGroup?, checkedId: Int ->
            when (checkedId) {
                R.id.chipLow -> viewModel.setPriority(
                    Priority.LOW
                )
                R.id.chipNormal -> viewModel.setPriority(Priority.NORMAL)
                R.id.chipHigh -> viewModel.setPriority(Priority.HIGH)
                else -> viewModel.setPriority(Priority.LOW)
            }
        }*/

        binding.chipGP.setOnCheckedStateChangeListener { _, checkedIds ->
            when (checkedIds[0]) {
                R.id.chipLow -> viewModel.priority = Priority.LOW

                R.id.chipNormal -> viewModel.priority = Priority.NORMAL

                R.id.chipHigh -> viewModel.priority = Priority.HIGH

                else -> viewModel.priority = Priority.LOW
            }
        }

        binding.inpEdtTitle.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun afterTextChanged(editable: Editable) {
                if (editable.isNotEmpty())
                    binding.inpTitle.error = null
            }
        })


        binding.mCardReminder.setOnClickListener {
            handlePickers(viewModel.oldDateTime?.date)
        }

        binding.aImgClear.setOnClickListener { viewModel.releaseAll() }

        binding.mBtnAdd.setOnClickListener {
            binding.inpTitle.error = null

            if (viewModel.isEditMode) {
                //edit mode
                val editedTodo = viewModel.editTodo(binding.inpEdtTitle.text.toString().trim())

                val res = todoViewModel.validateTodo(editedTodo)

                if (res == null) {
                    if (editedTodo.arriveDate != 0L)
                        notificationViewModel.setNotificationEditMode(editedTodo)
                    else  //if date not set or date cleared with imgClear
                        notificationViewModel.cancelAlarm(editedTodo)

                    todoViewModel.editTodo(editedTodo)

                    //searchViewModel.search()

                    updateDetail(editedTodo)

                    toastUtil.successToast(provideResource.getString(R.string.todo_edited_successfully_simple))

                    back()

                    return@setOnClickListener
                }

                binding.inpTitle.error = res

            } else {
                //add mode
                val newTodo = viewModel.addTodo(binding.inpEdtTitle.text.toString().trim())

                val res = todoViewModel.validateTodo(newTodo)

                if (res == null) {
                    todoViewModel.goToTop()

                    lifecycleScope.launch {
                        val insertedId = todoViewModel.addTodo(newTodo)

                        if (newTodo.arriveDate != 0L) {
                            newTodo.id = insertedId.toInt()
                            notificationViewModel.addNotification(newTodo)
                        }
                    }

                    toastUtil.successToast(provideResource.getString(R.string.todo_added_successfully_simple))

                    back()

                    if (viewModel.isShareMode) {
                        binding.mBtnAdd.isEnabled = false

                        val homeFragment: Fragment = HomeFragment.newInstance()

                        parentFragmentManager.beginTransaction().apply {
                            add(R.id.mainContainer, homeFragment, Constants.FragmentTag.HOME)
                        }.commit()
                    }


                    return@setOnClickListener
                }

                binding.inpTitle.error = res
            }
        }
    }

    private fun handlePickers(persianDate: PersianPickerDate?) {
        val picker = PersianDatePickerDialog(iContext)

        picker.setPositiveButtonString("مرحله بعد")
            .setNegativeButton("انصراف")
            .setTodayButton("تاریخ امروز")
            .setTodayButtonVisible(true)
            .setMinYear(1400)
            .setMaxYear(1440)
            .setTypeFace(Typeface.createFromAsset(iContext.assets, "font/vazir_rd_fd_medium.ttf"))
            .setTitleColor(provideResource.getColor(R.color.blue))
            .setAllButtonsTextSize(15)
            .setPickerBackgroundColor(provideResource.getColor(R.color.white))
            .setTitleType(PersianDatePickerDialog.WEEKDAY_DAY_MONTH_YEAR)
            .setShowInBottomSheet(true)
            .setEnableBellView(false)
            .setNotificationNoteClick { }
            .setListener(object : PersianPickerListener {
                override fun onDateSelected(
                    persianPickerDate: PersianPickerDate,
                    isPassed: Boolean
                ) {
                    if (isPassed) {
                        WarningDateDialog(iContext).apply {
                            show()

                            setTitle(provideResource.getString(R.string.passedDateTitle))

                            setMessage(provideResource.getString(R.string.passedDateMessage))

                            setEditText(provideResource.getString(R.string.editDate))

                            continueClicked = {
                                dismiss()
                                picker.dismiss()
                                dateSelectedAction(persianPickerDate)
                            }
                        }

                        return
                    }

                    picker.dismiss()
                    dateSelectedAction(persianPickerDate)
                }

                override fun onDismissed() {}
            })

        //init default values
        val date = viewModel.setInitDateValue(persianDate)

        val year = date[0]
        val month = date[1]
        val day = date[2]

        picker.setInitDate(year, month, day)
        picker.show()
    }

    private fun dateSelectedAction(persianPickerDate: PersianPickerDate?) {
        if (persianPickerDate == null)
            return

        viewModel.configTempDateTime()

        viewModel.setDateTemp(persianPickerDate)

        TimePickerSheetDialog(iContext, true, viewModel.tempDateTime).apply timerPicker@{
            show()

            onClickApply = applyClick@{ pickedDateTime: DateTime? ->
                if (viewModel.todayTtimePassed(pickedDateTime)) {

                    WarningDateDialog(iContext).apply {
                        setTitle(provideResource.getString(R.string.passedDateTitle))

                        setMessage(provideResource.getString(R.string.passedTimeMessage))

                        setEditText(provideResource.getString(R.string.editTime))

                        continueClicked = {
                            dismiss()
                            this@timerPicker.dismiss()

                            viewModel.commitDateTime(pickedDateTime)
                            viewModel.commitOldDateTime(pickedDateTime)
                        }

                        show()
                    }

                    return@applyClick
                }

                viewModel.commitDateTime(pickedDateTime)
                viewModel.commitOldDateTime(pickedDateTime)

                dismiss()
            }

            onBackClick = { date: PersianPickerDate? ->
                handlePickers(date)
            }
        }
    }

    private fun handleObserver() {
        collectLatestLifecycleFlow(viewModel.dateTimeStateFlow) { changedDateTime: DateTime? ->

            if (changedDateTime?.date != null) {
                binding.txtDate.setTextColor(
                    provideResource.getColor(R.color.black)
                )

                binding.txtDate.text = MessageFormat.format(
                    "{0}\nساعت {1}",
                    changedDateTime.persianDate,
                    changedDateTime.clock
                )

                binding.aImgClear.visibility = View.VISIBLE
            } else {
                binding.txtDate.setTextColor(
                    provideResource.getColor(R.color.gray)
                )

                binding.txtDate.text = provideResource.getString(R.string.set_date_time)
                binding.aImgClear.visibility = View.GONE
            }
        }

        collectLatestLifecycleFlow(viewModel.categoryStateFlow) {
            binding.txtCategory.text = viewModel.categoryTitleText

            if (viewModel.categoryIsValid())
                binding.txtCategory.setTextColor(provideResource.getColor(R.color.black))
            else
                binding.txtCategory.setTextColor(provideResource.getColor(R.color.gray))
        }

        collectLifecycleFlow(categoryViewModel.categoriesFlow) { categories ->
            allCategories = categories as ArrayList<Category>?
        }
    }

    private fun updateDetail(todo: Todo) {
        val fragment =
            parentFragmentManager.findFragmentByTag(Constants.FragmentTag.TODO_DETAIL) as TodoDetailFragment?

        fragment?.let {
            it.viewModel.todo = todo
        }
    }

}